src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java

```diff
@@ -181,31 +206,90 @@ private static void verifyLabel(Object label) {
      *     ...
      * }
      */
-    private static MethodHandle createRepeatIndexSwitch(MethodHandles.Lookup lookup, Object[] labels) {
-        MethodHandle def = MethodHandles.dropArguments(MethodHandles.constant(int.class, labels.length), 0, Object.class);
+    private static MethodHandle createRepeatIndexSwitch(MethodHandles.Lookup lookup, Object[] labels, Class<?> selectorType) {
+        MethodHandle def = MethodHandles.dropArguments(MethodHandles.constant(int.class, labels.length), 0, selectorType);
         MethodHandle[] testChains = new MethodHandle[labels.length];
         List<Object> labelsList = List.of(labels).reversed();
-
+        // unconditionally exact patterns always match
+        MethodHandle trueDef = MethodHandles.dropArguments(MethodHandles.constant(boolean.class, true), 0, selectorType, Object.class);
         for (int i = 0; i < labels.length; i++) {
             MethodHandle test = def;
             int idx = labels.length - 1;
             List<Object> currentLabels = labelsList.subList(0, labels.length - i);
-
             for (int j = 0; j < currentLabels.size(); j++, idx--) {
                 Object currentLabel = currentLabels.get(j);
+                Object testLabel = currentLabel;
                 if (j + 1 < currentLabels.size() && currentLabels.get(j + 1) == currentLabel) continue;
                 MethodHandle currentTest;
-                if (currentLabel instanceof Class<?>) {
-                    currentTest = INSTANCEOF_CHECK;
-                } else if (currentLabel instanceof Integer) {
-                    currentTest = INTEGER_EQ_CHECK;
-                } else if (currentLabel instanceof EnumDesc) {
+                if (currentLabel instanceof Class<?> currentLabelClass) {
+                    if (unconditionalExactnessCheck(selectorType, currentLabelClass)) {
+                        currentTest = trueDef;
+                    } else if (currentLabelClass.isPrimitive()) {
+                         if (selectorType.isInstance(Object.class)) {
```

Comment:
As discussed offline, I believe this is wrong, and it should be replaced by `Class::equals`. That said, I think that "accidentally" you get the desired behavior here, as demonstrated by jshell:

```
jshell> Byte.class.isInstance(Object.class)
$6 ==> false
```
```
jshell> Object.class.isInstance(Object.class)
$8 ==> true
```

So, this effectively acts as a test to check if the selector type is `Object`. Of course, since `isInstance` is used, spurious stuff is picked up as well:

```
jshell> Class.class.isInstance(Object.class)
$7 ==> true
```
But this situation turns out to be non problematic, given that a primitive pattern is not applicable to a selector type Class.

Reply:
Fixed as well.